feat(x-goog-spanner-request-id): implement Request-ID

odeke-em opened the pull request:
Implements x-goog-spanner-request-id propagation for end-to-end debugging.

Fixes googleapis/google-cloud-python#15905

odeke-em committed:
feat(x-goog-spanner-request-id): implement request_id generation and propagation

Generates a request_id that is then injected inside metadata
that's sent over to the Cloud Spanner backend.

Officially inject the first set of x-goog-spanner-request-id values into header metadata
Add request-id interceptor to use in asserting tests
Wrap Snapshot methods with x-goog-request-id metadata injector
Setup scaffolding for XGoogRequestIdHeader checks
Wire up XGoogSpannerRequestIdInterceptor for TestDatabase checks
Inject header in more Session using spots plus more tests
Base for tests with retries on abort
More plumbing for Transaction and Database
Update unit tests for Transaction
Wrap more in Transaction + update tests
Update tests
Plumb in more tests
Update TestDatabase

Fixes #1261

odeke-em committed:
More plumbing for Database DDL methods

odeke-em committed:
Update test_spannner.test_transaction tests

odeke-em committed:
Update test_session tests

olavloite commented on google/cloud/spanner_v1/database.py:
```diff
@@ -714,17 +775,24 @@ def execute_pdml():
                     request=request,
                     transaction_selector=txn_selector,
                     observability_options=self.observability_options,
+                    attempt=begin_txn_attempt,
```
I think that this should be `partial_attempt`. Retries of this RPC will not retry the `BeginTransaction` RPC, it will just continue to use the same transaction ID as the one that was returned by the original `BeginTransaction` RPC.

odeke-em replied:
Oooh, okay, that's very interesting and thanks for letting me know, let me make that update.

odeke-em replied:
I did however already pass in and increment partial_attempt inside wrapped_method. I'll instead remove that and increment them in the appropriate sites.

olavloite commented on google/cloud/spanner_v1/instance.py:
```diff
@@ -501,6 +501,7 @@ def database(
                 proto_descriptors=proto_descriptors,
             )
         else:
+            print("enabled interceptors")
```
nit: remove?

olavloite commented on google/cloud/spanner_v1/pool.py:
```diff
@@ -243,6 +243,7 @@ def bind(self, database):
             "CloudSpanner.FixedPool.BatchCreateSessions",
             observability_options=observability_options,
         ) as span:
+            attempt = 1
```
nit: I don't think we need this. We can just pass in 1 as a constant in the function below just as for the other functions.

olavloite commented on google/cloud/spanner_v1/snapshot.py:
```diff
@@ -92,6 +95,7 @@ def _restart_on_unavailable(
     ):
         iterator = method(request=request)
     while True:
+        attempt += 1
```
I think that this should be moved into the `except ServiceUnavailable` block. The logic in this function is:
1. Iterate until you encounter a `resume_token`. When we see that, then yield what is in the buffer, as we know that it will be safe to retry.
2. If a `ServiceUnavailable` error is encountered, then do another call (a retry) and replace the underlying iterator with the one that is returned by that new call.

olavloite replied:
It should also be incremented in case of an `InternalServerError` that is retryable.

olavloite replied:
Or can it be removed altogether? It does not seem to be used?

odeke-em replied:
It was being used as a side effect but sure I've removed it and performing the increment appropriately where directly used.

olavloite commented on google/cloud/spanner_v1/testing/interceptors.py:
```diff
@@ -63,3 +65,75 @@ def reset(self):
+
+        response_or_iterator = method(request_or_iterator, call_details)
+        streaming = getattr(response_or_iterator, "__iter__", None) is not None
+        print(
```
nit: remove?

olavloite commented on tests/unit/test_request_id_header.py:
```diff
@@ -0,0 +1,387 @@
+            if n_finished == len(threads):
+                break
+
+            time.sleep(1)
```
nit: can we remove this? The call to `thread.join()` should ensure that the threads have finished.

olavloite commented on tests/unit/test_request_id_header.py:
```diff
@@ -0,0 +1,387 @@
+                (1, REQ_RAND_PROCESS_ID, client_id, channel_id, 1, 1),
+            ),
+            (
+                "/google.spanner.v1.Spanner/GetSession",
```
Note: [This PR](https://github.com/googleapis/python-spanner/pull/1252) should get rid of all of these `GetSession` calls, so rebasing on main my cause this test to fail.

odeke-em replied:
Roger that and thanks!

olavloite commented on tests/unit/test_request_id_header.py:
```diff
@@ -0,0 +1,387 @@
+        ]
+        assert got_stream_segments == want_stream_segments
+
+    def test_database_run_in_transaction_retries_on_abort(self):
```
This test does not appear to actually verify anything (other than that it does not cause any errors). Can we either remove it, or add verifications for the scenario it should cover?

olavloite commented on tests/unit/test_session.py:
```diff
@@ -1113,10 +1174,25 @@ def unit_of_work(txn, *args, **kw):
+                        ("x-goog-spanner-route-to-leader", "true"),
+                        (
+                            "x-goog-spanner-request-id",
+                            f"1.{REQ_RAND_PROCESS_ID}.{database.NTH_CLIENT.value}.1.2.2",
```
Is `attempt` set to 2 here because the transaction is retried? If so, then that is not correct. The `attempt` number should only be increased if the RPC itself is being retried due to a retryable error for that (single) RPC. E.g. if a `ServiceUnavailable` error is returned.

olavloite commented on tests/unit/test_session.py:
```diff
@@ -1132,10 +1208,24 @@ def unit_of_work(txn, *args, **kw):
+                        ("x-goog-spanner-route-to-leader", "true"),
+                        (
+                            "x-goog-spanner-request-id",
+                            f"1.{REQ_RAND_PROCESS_ID}.{database.NTH_CLIENT.value}.1.3.2",
```
Same here as for the `BeginTransaction` RPC: Transaction retries should not increase the attempt count. Only retries of a single RPC due to `UNAVAILABLE` errors (and other retryable error codes).

odeke-em commented on google/cloud/spanner_v1/database.py:
```diff
@@ -443,6 +448,31 @@ def spanner_api(self):
+                channel_id = len(self.__transports_to_channel_id) + 1
+                self.__transports_to_channel_id[api._transport] = channel_id
+
+            return channel_id
```
It is indeed cached as it is saved in the Database static __transports_to_channel_id map or would you prefer I name it with capital letters so that it is much easier to spot as __TRANSPORTS_TO_CHANNEL_ID?

olavloite replied:
Yeah, what I meant is that I think we can just add an instance field `self.__channel_id` or something like that and store it there. Now it is indeed cached, but the cache look up includes taking a lock and then looking it up in a map. This is a function that is called for __every__ RPC, so keeping it as quick as possible is desirable, and I expect the extra memory usage to be negligible, as I don't think any application is going to create a very large number of database objects.

odeke-em replied:
Gotcha and thank you @olavloite, done!

odeke-em commented on google/cloud/spanner_v1/database.py:
```diff
@@ -698,10 +728,20 @@ def execute_partitioned_dml(
                 _metadata_with_leader_aware_routing(self._route_to_leader_enabled)
             )
 
+        # Attempt will be incremented inside _restart_on_unavailable.
```
It applies to `begin_txn_attempt` and I've updated the comment and moving the comment before the variable itself. Thank you.

odeke-em committed:
Update tests

odeke-em committed:
Update propagation of changes

odeke-em committed:
Plumb test_context_manager

odeke-em committed:
Fix more tests

odeke-em committed:
More test plumbing

odeke-em committed:
Infer database._channel_id only once along with spanner_api

odeke-em committed:
Update batch tests

olavloite committed:
test: add tests for retries and move mock server test to correct directory

This commit contains a couple of changes:
1. Moves the request-ID tests using the mock server to the mockserver tests directory.
2. Adds tests for retries of RPCs. These currently fail, as the request ID is not
   updated correctly when a retry is executed for a unary RPC. Also, the retry test
   for streaming RPCs fails, but this is due to an existing bug in the client lib.

olavloite committed:
fix: revert default Python version

odeke-em committed:
Fix discrepancy with api.batch_create_sessions automatically retrying assuming idempotency with headers too

odeke-em committed:
Take into account current behavior of /GetSession /BatchCreateSession in tests

odeke-em commented:
Kindly help me take another pass @olavloite. Thank you and happy New Year to you!

odeke-em committed:
Implement interceptor to wrap and increase x-goog-spanner-request-id attempts per retry

This monkey patches SpannerClient methods to have an interceptor that
increases the attempts per retry. The prelude though is that any
callers to it must pass in the attempt value 0 so that each pass through
will correctly increase the attempt field's value.

odeke-em committed:
Correctly handle wrapping by class for api objects

odeke-em committed:
Revert poool creation interception attempts

odeke-em committed:
Wire up and revert some prints

olavloite commented on tests/mockserver_tests/test_request_id_header.py:
```diff
@@ -0,0 +1,363 @@
+    def test_streaming_retryable_error(self):
+        add_select1_result()
+        add_error(SpannerServicer.ExecuteStreamingSql.__name__, unavailable_status())
+        add_error(SpannerServicer.ExecuteStreamingSql.__name__, unavailable_status())
```
nit: I think that adding one error is enough

odeke-em committed:
Initial ExecuteStreamingSql request in snapshot should have the header

odeke-em committed:
Consolidate retries for _restart_on_unavailable

odeke-em committed:
Fix missing variable declaration

odeke-em committed:
Adjust with updates

odeke-em committed:
Update _execute_partitioned_dml_helper

odeke-em committed:
Monkey patch updates

odeke-em committed:
Merge branch 'main' into x-goog-spanner-request-id

odeke-em committed:
Reduce unnecessary changes

odeke-em committed:
Experiment with wrapping for gapic retries

odeke-em committed:
Remove duplicate code

odeke-em committed:
Complete mockserver tests
